File: backend/src/main/java/com/team01/backend/domain/post/dto/PostDetailResponseDto.java
package com.team01.backend.domain.post.dto;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.team01.backend.domain.board.entity.Board;
import com.team01.backend.domain.category.entity.Category;
import com.team01.backend.domain.comment.dto.CommentReadResponseDto;
import com.team01.backend.domain.post.entity.Post;

import java.time.LocalDateTime;
import java.util.List;

public record PostDetailResponseDto(
        Long id,
        Long boardId,
        String boardName,
        Long categoryId,
        String categoryName,
        String title,
        String content,
        String author,
        String profileImage,
        int likeCount,
        boolean isLiked,

        @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
        LocalDateTime createdAt,
        @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
        LocalDateTime modifiedAt,

        List<CommentReadResponseDto> comments,
        boolean isOwner
) {
    public static PostDetailResponseDto of(Post post, Board board, Category category,
                                           List<CommentReadResponseDto> comments, boolean isOwner, boolean isLiked) {
        return new PostDetailResponseDto(
                post.getId(),
                board.getId(),
                board.getName(),
                category.getId(),
                category.getName(),
                post.getTitle(),
                post.getContent(),
                post.getAuthor().getNickname(),
                post.getAuthor().getProfileImage(),
                post.getLikeCount(),
                isLiked,
                post.getCreatedAt(),
                post.getModifiedAt(),
                comments,
                isOwner
        );
    }
}

File: backend/src/main/java/com/team01/backend/domain/post/dto/PostDetailResponseDto.kt
package com.team01.backend.domain.post.dto

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonProperty
import com.team01.backend.domain.board.entity.Board
import com.team01.backend.domain.category.entity.Category
import com.team01.backend.domain.comment.dto.CommentReadResponseDto
import com.team01.backend.domain.post.entity.Post

import java.time.LocalDateTime

data class PostDetailResponseDto(
        val id: Long,
        val boardId: Long,
        val boardName: String,
        val categoryId: Long,
        val categoryName: String,
        val title: String,
        val content: String,
        val author: String,
        val profileImage: String?,
        val likeCount: Int,
        @get:JsonProperty("isLiked")
        val liked: Boolean,
        @field:JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
        val createdAt: LocalDateTime,
        @field:JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
        val modifiedAt: LocalDateTime,
        val comments: List<CommentReadResponseDto>,
        @get:JsonProperty("isOwner")
        val owner: Boolean,
) {
        companion object {
                @JvmStatic
                fun of(
                        post: Post,
                        board: Board,
                        category: Category,
                        comments: List<CommentReadResponseDto>,
                        isOwner: Boolean,
                        isLiked: Boolean,
                ): PostDetailResponseDto = PostDetailResponseDto(
                        id = post.id ?: throw IllegalStateException("Post id is null"),
                        boardId = board.id ?: throw IllegalStateException("Board id is null"),
                        boardName = board.name,
                        categoryId = category.id ?: throw IllegalStateException("Category id is null"),
                        categoryName = category.name,
                        title = post.title,
                        content = post.content,
                        author = post.author.nickname,
                        profileImage = post.author.profileImage,
                        likeCount = post.likeCount,
                        liked = isLiked,
                        createdAt = post.createdAt,
                        modifiedAt = post.modifiedAt,
                        comments = comments,
                        owner = isOwner,
                )
        }
}
